Initial implementation of nanmean

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62671

Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

fixes https://github.com/pytorch/pytorch/issues/21987

**TODO**

- [x] Fix autograd issue

Differential Revision: [D30515181](https://our.internmc.facebook.com/intern/diff/D30515181)

heitorschueroff committed:
Initial implementation of nanmean

[ghstack-poisoned]

heitorschueroff committed:
Update on "Initial implementation of nanmean"




[ghstack-poisoned]

mruberry commented:
Doc build failure is real:

```
Aug 03 22:37:19 /opt/conda/lib/python3.6/site-packages/torch/__init__.py:docstring of torch.nanmean:40: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```

those build failures are still very hard to read, unfortunately

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1036,6 +1036,80 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
+      dim_prod *= self.size(d);
+    }
+  }
+  const auto factor = dim_prod - self.isnan().sum(dim, keepdim);
```
Add a comment here for the algorithm

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1036,6 +1036,80 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
+    IntArrayRef dim,
+    bool keepdim,
+    ScalarType dtype) {
+  int64_t dim_prod = 1;
```
Comment here for what this section does

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1036,6 +1036,80 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
+      opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(scalarType),
+      "Can only calculate the mean of floating types. Got ",
```
Start error messages with the name of the op, like "torch.nanmean: ..."

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1036,6 +1036,80 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
+      toString(scalarType),
+      " instead.");
+  ScalarType dtype = get_dtype_from_result(result, opt_dtype);
+  result.set_(nanmean_impl(self, dim, keepdim, dtype));
```
I don't think setting result is correct -- don't the OpInfo out= tests complain about this? Like isn't this just throwing away pre-allocated storages? Or how would this work if result previously had an incorrect dtype, or was a view of a larger tensor?

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1036,6 +1036,80 @@ Tensor& mean_out(const Tensor& self, DimnameList dim,
+  return result;
+}
+
+Tensor& nanmean_out(
```
I wouldn't support named tensors

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2815,6 +2815,24 @@
   device_check: NoCheck   # TensorIterator
 
+- func: nanmean(Tensor self, *, ScalarType? dtype=None) -> Tensor
+  device_check: NoCheck   # TensorIterator
```
"#TensorIterator" seems misleading here because nanmean is a composite operation (although it's true that the ops it calls are implemented using TensorIterator). It's probably more accurate to comment this as a "#Composite"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
 """.format(**multi_dim_common))
 
+add_docstr(torch.nanmean, r"""
+nanmean(input) -> Tensor
```
Can we document this signature in one function? The double signature thing is lame and an artifact of imperfect op implementations

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
+
+.. note::
+
+    `torch.nanmean(a)` is equivalent to `torch.mean(a[~a.isnan()])`
```
I would make this a parenthetical after the discussion of "identical" above

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
+
+    `torch.nanmean(a)` is equivalent to `torch.mean(a[~a.isnan()])`
+
+Args:
```
Add a seealso section for mean

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
+.. function:: mean(input, dim, keepdim=False, *, out=None) -> Tensor
+   :noindex:
+
+Computes the mean of all `non-NaN` elements alogn the specified :attr:`dim` in
```
"alogn" -> "along"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> x = torch.tensor([[torch.nan, 1, 2], [1, 2, 3]])
```
Good example showing the difference between mean and nanmean

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6010,6 +6009,60 @@ def merge_dicts(*dicts):
+    >>> x.nanmean(dim=0)
+    tensor([1.0000, 1.5000, 2.5000])
+
+    # If all elements in the reduced dimensions are NaN then the result is NaN
```
Nice edge case clarification

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2269,6 +2275,23 @@ def fn(op_info, device, dtype, requires_grad):
 
     return fn
 
+def sample_inputs_nan_reduction(supports_multiple_dims):
```
Seems like this should either combine the above and the normal inputs (to verify it handles non-nan inputs properly) or the above nan inputs should be part of the regular inputs to every reduction operator and the nan ops shouldn't have a special-cased sample inputs

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6454,6 +6477,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           # Need to skip out test because one of the overload for mean does not support it
+           # TODO(@heitorschueroff) fix this when implementing ReductionInfo
+           skips=(
+               SkipInfo('TestCommon', 'test_out'),
```
Per the above comment, I think it's important we investigate the set_ behavior, make sure we're clear on whether it's OK for out= or not OK and why, and that test_out can detect that behavior if we decide it's not OK -- so I don't think we should skip test_out. Maybe that means waiting for some ReductionInfo PRs

mruberry reviewed (commented):
Overall looks good but some investigation required (see comments inline) and a few suggestions/questions for your review

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

**TODO**

- [ ] Fix autograd issue

[ghstack-poisoned]

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

**TODO**

- [ ] Fix autograd issue

[ghstack-poisoned]

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

**TODO**

- [x] Fix autograd issue

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30515181).

mruberry reviewed (approved):
Nice work! Just a couple minor inline comments for your review

TODO: make this a ReductionInfo for more thorough testing

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

**TODO**

- [x] Fix autograd issue

Differential Revision: [D30515181](https://our.internmc.facebook.com/intern/diff/D30515181)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

fixes https://github.com/pytorch/pytorch/issues/21987

**TODO**

- [x] Fix autograd issue

Differential Revision: [D30515181](https://our.internmc.facebook.com/intern/diff/D30515181)

[ghstack-poisoned]

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -2421,13 +2421,18 @@ def callable(a, b) -> number
-               r"""
-mean(dim=None, keepdim=False) -> Tensor or (Tensor, Tensor)
+add_docstr_all('mean', r"""
+mean(dim=None, keepdim=False) -> Tensor
```
Nice correction

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5942,8 +5942,7 @@ def merge_dicts(*dicts):
-add_docstr(torch.mean,
-           r"""
+add_docstr(torch.mean, r"""
 mean(input) -> Tensor
```
Weirdly I think this signature is `(input, *, dtype)`?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5974,8 +5973,13 @@ def merge_dicts(*dicts):
+    {dtype}
     {out}
 
+.. seealso::
```
Nice see also

mruberry reviewed (approved):
Cool! I made a few small comments about fixing the signature of mean still. Overall looks good.

There is also one test that still needs to be addressed: test_ref_duplicate_values_nanmean_cuda_float16.

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

fixes https://github.com/pytorch/pytorch/issues/21987

**TODO**

- [x] Fix autograd issue

Differential Revision: [D30515181](https://our.internmc.facebook.com/intern/diff/D30515181)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Initial implementation of nanmean"


Very crude first implementation of `torch.nanmean`. The current reduction kernels do not have good support for implementing nan* variants. Rather than implementing new kernels for each nan* operator, I will work on new reduction kernels with support for a `nan_policy` flag and then I will port `nanmean` to use that.

fixes https://github.com/pytorch/pytorch/issues/21987

**TODO**

- [x] Fix autograd issue

Differential Revision: [D30515181](https://our.internmc.facebook.com/intern/diff/D30515181)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30515181).
